tlaplus/ghostferry.tla

```diff
@@ -224,25 +222,92 @@ NoRecordHere == CHOOSE r : r \notin Records
 \* A set of possible records for TypeOK
 PossibleRecords == Records \cup {NoRecordHere}
 
+\* These are all the possible states of the table, but they may not be valid.
+PossibleTableStates == [1..TableCapacity -> PossibleRecords]
+
+\* This is a set of all possible binlog structs in a partial form: it is missing the pk.
+PossibleOldNewRowStates == {s \in [{"oldr", "newr"} -> Records] : s.oldr # s.newr}
+
+\* This is an operator that adds the pk back into a partial binlog struct in PossibleOldNewRowStates and adds back the PK.
+ConstructBinlogFromOldNewRowStates(k, s) == [pk |-> k] @@ s
+
+\* This generates a set of possible binlog entries with all possible PKs.
+PossibleBinlogEntries == UNION {{ConstructBinlogFromOldNewRowStates(k, s) : k \in PrimaryKeys} : s \in PossibleOldNewRowStates}
+
+\* This generates a set of possible SourceBinlog, which is a sequence.
+\* For simplicity and TLC feasibility, we only generate a single binlog entry in the initial possible SourceBinlog.
```

Comment:
Hmm, would it make sense to generate the number of entries up to the `LastSuccessfulPK ` instead of just a single entry? I do know we are simplifying 1 == n here and in other places, so maybe that doesn't make sense.

Reply:
It would work I think, but it would explode the initial states. Currently already taking 20min to check on ryzen 3700x so if we introduce more states we will likely need to state check in a much more powerful machine...